homeassistant/components/notify/__init__.py

```diff
@@ -4,110 +4,281 @@
+            (ATTR_LONGITUDE, longitude),
+            (ATTR_DEVICE, device_entity),
+            (ATTR_CAPTION, caption)
+        ] if value is not None
+    }
+
+    hass.services.call(DOMAIN, SERVICE_SEND_LOCATION, data)
 
 
 def setup(hass, config):
     """Setup the notify services."""
-    success = False
+    component = EntityComponent(
+        logging.getLogger(__name__), DOMAIN, hass)
+
+    component.setup(config)
 
     descriptions = load_yaml_config_file(
         os.path.join(os.path.dirname(__file__), 'services.yaml'))
 
-    for platform, p_config in config_per_platform(config, DOMAIN):
-        notify_implementation = bootstrap.prepare_setup_platform(
-            hass, config, DOMAIN, platform)
+    def message_handle_service(service):
+        """Handle sending notification message service calls."""
+        message = service.data.get(ATTR_MESSAGE)
 
-        if notify_implementation is None:
-            _LOGGER.error("Unknown notification service specified.")
-            continue
+        title = template.render(
+            hass, service.data.get(ATTR_TITLE, ATTR_TITLE_DEFAULT))
+        target = service.data.get(ATTR_TARGET)
+        message = template.render(hass, message)
+        formats = service.data.get(ATTR_FORMATS)
 
-        notify_service = notify_implementation.get_service(hass, p_config)
+        for notify in component.extract_from_service(service):
+            notify.send_message(message, title=title, target=target,
+                                formats=formats)
 
-        if notify_service is None:
-            _LOGGER.error("Failed to initialize notification service %s",
-                          platform)
-            continue
+    hass.services.register(DOMAIN, SERVICE_SEND_MESSAGE,
+                           message_handle_service,
+                           descriptions.get(SERVICE_SEND_MESSAGE),
+                           schema=MESSAGE_SERVICE_SCHEMA)
 
-        def notify_message(notify_service, call):
-            """Handle sending notification message service calls."""
-            message = call.data[ATTR_MESSAGE]
+    def photo_handle_service(service):
+        """Handle sending photo service calls."""
+        caption = service.data.get(ATTR_CAPTION, None)
+        if caption is not None:
+            caption = template.render(hass, caption)
 
-            title = template.render(
-                hass, call.data.get(ATTR_TITLE, ATTR_TITLE_DEFAULT))
-            target = call.data.get(ATTR_TARGET)
-            message = template.render(hass, message)
-            data = call.data.get(ATTR_DATA)
+        photo = service.data.copy()
+        photo.pop(ATTR_CAPTION, None)
+        photo.pop(ATTR_ENTITY_ID, None)
 
-            notify_service.send_message(message, title=title, target=target,
-                                        data=data)
+        for notify in component.extract_from_service(service):
+            notify.send_photo(photo, caption=caption)
```

Comment:
Resolving the photo into a file / stream / bytes should happen in the service handler. Otherwise each platform will replicate logic